### PR TITLE
Implemented AddQuantDequantPass in imperative quantization.

### PR DESCRIPTION
**Pull request by huangxu96**
### PR types

New features
### PR changes

APIs
### Describe

Finished API for imperative quantization. Now it can support the AddQuantDequantPass in imperative quantization.

AddQuantDequantPass allows users to quantize noweight layers in model like pool2d and ReLU.

**Usage:** Passing the layers name you want to quantize when you use ImperativeQuantAware API.

e.g. 

```python
        imperative_qat = ImperativeQuantAware(
            weight_quantize_type='abs_max',
            activation_quantize_type='moving_average_abs_max',
            quantizable_layer_type=[
                'Conv2D', 'Linear', 'ReLU', 'Pool2D', 'LeakyReLU', 'ReLU6',
                'Tanh', 'Swish'
            ]) 

```

Supported noweight layers:

`python ['ReLU', 'Pool2D', 'LeakyReLU', 'ReLU6', 'Tanh', 'Swish']`

We are continue supporting more layers.

The accuracy comparison are as follows：

image

### CODE REVIEW

File: python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py [hunk 1]
```diff
@@ -494,3 +495,27 @@ def forward(self, input):
...
+    def __init__(self,
+                 layer,
+                 weight_bits=8,
+                 activation_bits=8,
+                 moving_rate=0.9,
+                 weight_quantize_type='abs_max',
+                 activation_quantize_type='abs_max'):
+
+        super(QuantizedNoweightLayer, self).__init__()
+        self.layer = layer
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py]: Please use `self._layer` instead.

File: python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py [hunk 2]
```diff
@@ -494,3 +495,27 @@ def forward(self, input):
...
+
+
+class QuantizedNoweightLayer(layers.Layer):
+    def __init__(self,
+                 layer,
+                 weight_bits=8,
+                 activation_bits=8,
+                 moving_rate=0.9,
+                 weight_quantize_type='abs_max',
+                 activation_quantize_type='abs_max'):
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py]: According to static quantization, this should always be `moving_average_abs_max`. Just refine as follow,

```python
    def __init__(self,
                 layer,
                 weight_bits=8,
                 activation_bits=8,
                 moving_rate=0.9,
                *args,
                **kwargs):
```

File: python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py [hunk 3]
```diff
@@ -494,3 +495,27 @@ def forward(self, input):
...
+
+        super(QuantizedNoweightLayer, self).__init__()
+        self.layer = layer
+        self._fake_quant_input = _get_fake_quant_type(
+            activation_quantize_type,
+            name=layer.full_name(),
+            moving_rate=moving_rate,
+            quant_bits=activation_bits,
+            dtype=self._dtype,
+            quant_on_weight=False)
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py]: Just refine as follow,
```python
        self._fake_quant_input = _get_fake_quant_type(
            'moving_average_abs_max',
            name=layer.full_name(),
            moving_rate=moving_rate,
            quant_bits=activation_bits,
            dtype=self._dtype,
            quant_on_weight=False)
```

File: python/paddle/fluid/contrib/slim/tests/test_imperative_qat_AddQuantDequant.py [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+    fc_b3_attr = fluid.ParamAttr(name="fc_b_3")
+    conv1 = fluid.layers.conv2d(
+        data,
+        num_filters=6,
+        filter_size=3,
+        stride=1,
+        padding=1,
+        param_attr=conv2d_w1_attr,
+        bias_attr=conv2d_b1_attr)
+    conv1 = paddle.fluid.layers.leaky_relu(conv1)
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/tests/test_imperative_qat_AddQuantDequant.py]:
```python
conv1 = paddle.fluid.layers.leaky_relu(conv1,  alpha=0.02)
```

File: python/paddle/fluid/contrib/slim/tests/test_imperative_qat_AddQuantDequant.py [hunk 5]
```diff
@@ -0,0 +1,445 @@
...
+            data = next(test_reader())
+            test_data = np.array([x[0].reshape(1, 28, 28)
+                                  for x in data]).astype('float32')
+            test_img = fluid.dygraph.to_variable(test_data)
+            lenet.eval()
+            before_save = lenet(test_img)
+
+        # save inference quantized model
+        path = "./mnist_infer_model"
+        paddle.jit.save(
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/tests/test_imperative_qat_AddQuantDequant.py]: Use `paddle.jit.save(lenet.eval(), ...)` instead.

File: python/paddle/fluid/contrib/slim/tests/test_imperative_qat_AddQuantDequant.py [hunk 6]
```diff
@@ -0,0 +1,445 @@
...
+                avg_loss = fluid.layers.mean(loss)
+                avg_loss.backward()
+                adam.minimize(avg_loss)
+                lenet.clear_gradients()
+                dynamic_loss_rec.append(avg_loss.numpy()[0])
+                if batch_id % 100 == 0:
+                    _logger.info('{}: {}'.format('loss', avg_loss.numpy()))
+
+        paddle.jit.save(
+            layer=lenet,
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/tests/test_imperative_qat_AddQuantDequant.py]: Use `paddle.jit.save(lenet.eval(),...)` instead.

**Review by wzzju — CHANGES_REQUESTED** [review]

File: python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py (hunk 1 above)
Comment (huangxu96) [reply python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py]: Done.

File: python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py (hunk 2 above)
Comment (huangxu96) [reply python/paddle/fluid/contrib/slim/quantization/imperative/quant_nn.py]: Done.

File: python/paddle/fluid/contrib/slim/tests/test_imperative_qat_AddQuantDequant.py (hunk 4 above)
Comment (huangxu96) [reply python/paddle/fluid/contrib/slim/tests/test_imperative_qat_AddQuantDequant.py]: Done

File: python/paddle/fluid/contrib/slim/tests/test_imperative_qat_AddQuantDequant.py (hunk 5 above)
Comment (huangxu96) [reply python/paddle/fluid/contrib/slim/tests/test_imperative_qat_AddQuantDequant.py]: Done

File: python/paddle/fluid/contrib/slim/tests/test_imperative_qat_AddQuantDequant.py (hunk 6 above)
Comment (huangxu96) [reply python/paddle/fluid/contrib/slim/tests/test_imperative_qat_AddQuantDequant.py]: Done.

**Review by juncaipeng — COMMENTED** [review]
Cannot quantize multiple input ops, such as elementwise_add

这个暂时不考虑，后面再支持

File: python/paddle/fluid/contrib/slim/quantization/imperative/qat.py [hunk 7]
```diff
@@ -78,7 +78,7 @@ def __init__(self,
                 'moving_average_abs_max', the static quantization scale will be calculated
                 during training and used in inference.
             moving_rate(float): the parameter for 'moving_average_abs_max' quantization.
-            quantizable_op_type(list[str]): List the type of layers that will be quantized. 
+            quantizable_layer_type(list[str]): List the type of layers that will be quantized.
```
Comment (juncaipeng) [inline python/paddle/fluid/contrib/slim/quantization/imperative/qat.py:89]: 参考pytorch和tf，建议这里用户传入paddle.nn.Conv2D等layer，不要传入str

File: python/paddle/fluid/contrib/slim/quantization/imperative/qat.py [hunk 8]
```diff
@@ -186,8 +196,16 @@ def _get_quantized_counterpart(self, layer):
...
                 layer.full_name()))
             sys.exit(-1)
-
-        quantized_layer = quant_nn.__dict__[quantized_counterpart[index]](
+        no_weight_layer = [
+            'QuantizedPool2D', 'QuantizedReLU', 'QuantizedLeakyReLU',
+            'QuantizedReLU6', 'QuantizedSoftmax', 'QuantizedTanh',
+            'QuantizedSwish'
+        ]
+        if quantized_counterpart[index] in no_weight_layer:
```
Comment (juncaipeng) [inline python/paddle/fluid/contrib/slim/quantization/imperative/qat.py]: 后续会有很多无权重op的量化，建议这个if判断调换一下顺序，判断是否是有权重量化op

Comment (huangxu96) [reply python/paddle/fluid/contrib/slim/quantization/imperative/qat.py]: Done.

File: python/paddle/fluid/contrib/slim/quantization/imperative/qat.py (hunk 7 above)
Comment (huangxu96) [reply python/paddle/fluid/contrib/slim/quantization/imperative/qat.py:89]: 下个PR 会考虑改成这总写法。

**Review by juncaipeng — DISMISSED** [review]
LGTM

**Review by XieYunshen — APPROVED** [review]

**Review by juncaipeng — APPROVED** [review]
LGTM

**Review by juncaipeng — APPROVED** [review]
LGTM

**Review by wzzju — APPROVED** [review]
LGTM.
